src/inference/mh.jl

```diff
@@ -187,34 +201,77 @@ function AbstractMCMC.sample_init!(
 
     # Get `init_theta`
     initialize_parameters!(spl; verbose=verbose, kwargs...)
+
+    # If we're doing random walk with a covariance matrix,
+    # just link everything before sampling.
+    if spl.alg.proposals isa AdvancedMH.RandomWalkProposal{<:MvNormal}
+        link!(spl.state.vi, spl)
+    end
 end
 
-function AbstractMCMC.step!(
+function AbstractMCMC.sample_end!(
     rng::AbstractRNG,
     model::Model,
     spl::Sampler{<:MH},
     N::Integer,
-    transition;
+    transitions;
+    verbose::Bool=true,
+    resume_from=nothing,
     kwargs...
 )
+    # We are doing a random walk, so we unlink everything when we're done.
+    if spl.alg.proposals isa AdvancedMH.RandomWalkProposal{<:MvNormal}
+        invlink!(spl.state.vi, spl)
+    end
+end
+
+function AbstractMCMC.step!(
+    rng::AbstractRNG,
+    model::Model,
+    spl::Sampler{MH{space, P}},
+    N::Integer,
+    transition;
+    kwargs...
+) where {space, P}
     if spl.selector.rerun # Recompute joint in logp
         model(spl.state.vi)
     end
 
-    # Retrieve distribution and value NamedTuples.
-    dt, vt = dist_val_tuple(spl)
-
-    # Create a sampler and the previous transition.
-    mh_sampler = AMH.MetropolisHastings(dt)
-    prev_trans = AMH.Transition(vt, getlogp(spl.state.vi))
-
-    # Make a new transition.
-    densitymodel = AMH.DensityModel(MHLogDensityFunction(model, spl))
-    trans = AbstractMCMC.step!(rng, densitymodel, mh_sampler, 1, prev_trans)
-
-    # Update the values in the VarInfo.
-    set_namedtuple!(spl.state.vi, trans.params)
-    setlogp!(spl.state.vi, trans.lp)
+    # Cases:
+    # 1. A covariance proposal matrix
+    # 2. A bunch of NamedTuples that specify the proposal space
+    if spl.alg.proposals isa AdvancedMH.RandomWalkProposal{<:MvNormal}
```

Comment:
Also here I think it would be natural to call into a method that allows to dispatch on the type of proposal instead of hardcoding the check in the general implementation of `step!`.

Reply:
Okay, I added a `propose!` function that does this.